fuzz/fuzz_targets/fuzz.rs

```diff
@@ -0,0 +1,320 @@
+            let traverse_iterator = bvh
+                .traverse_iterator(&ray, &self.shapes)
+                .map(ByPtr)
+                .collect::<HashSet<_>>();
+            let _traverse_flat = flat_bvh
+                .traverse(&ray, &self.shapes)
+                .into_iter()
+                .map(ByPtr)
+                .collect::<HashSet<_>>();
+
+            if assert_traversal_agreement {
+                assert_eq!(traverse, traverse_iterator);
+                // Fails, due to a bug.
+                // assert_eq!(traverse, traverse_flat);
+            } else {
+                // Fails, probably due to rounding errors.
+            }
+
+            let nearest_traverse_iterator = bvh
+                .nearest_traverse_iterator(&ray, &self.shapes)
+                .map(ByPtr)
+                .collect::<HashSet<_>>();
+            let farthest_traverse_iterator = bvh
+                .farthest_traverse_iterator(&ray, &self.shapes)
+                .map(ByPtr)
+                .collect::<HashSet<_>>();
+
+            if assert_traversal_agreement {
+                // Fails, due to a bug.
+                // assert_eq!(traverse_iterator, nearest_traverse_iterator);
+            } else {
+                // Fails, probably due to rounding errors.
+            }
+
+            // Since the algorithm is similar, these should agree regardless of mode.
+            assert_eq!(nearest_traverse_iterator, farthest_traverse_iterator);
+
+            if let Some(mutation) = self.mutations.pop() {
+                match mutation {
+                    ArbitraryMutation::Add(shape) => {
+                        let new_shape_index = self.shapes.len();
+                        self.shapes.push(shape);
+                        bvh.add_shape(&mut self.shapes, new_shape_index);
+                    }
+                    ArbitraryMutation::Remove(index) => {
+                        // TODO: remove `false &&` once this no longer causes a panic:
+                        // "Circular node that wasn't root parent=0 node=2"
+                        if false
+                        /* index < self.shapes.len() */
+                        {
+                            bvh.remove_shape(&mut self.shapes, index, true);
+                            self.shapes.pop().unwrap();
+                        }
+                    }
+                }
+            } else {
+                break;
+            }
+        }
+    }
+}
+
+/// Makes it easy to compare sets of intersected shapes. Comparing by
+/// value would be ambiguous if multiple shapes shared the same AABB.
+#[derive(Debug)]
+struct ByPtr<'a, T>(&'a T);
+
+impl<T> PartialEq for ByPtr<'_, T> {
+    fn eq(&self, other: &Self) -> bool {
+        std::ptr::eq(self.0, other.0)
+    }
+}
+
+impl<T> Eq for ByPtr<'_, T> {}
+
+impl<T> Hash for ByPtr<'_, T> {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        state.write_usize(self.0 as *const _ as usize);
+    }
+}
```

Comment:
This feels a bit like a crutch and it seems there should be a better way for us to compare these but I don't really have a better idea here either.

Reply:
This was the best thing I could think of. If this was a normal unit test, I'd give each shape a unique numerical ID and compare sets of ID's. However, an *arbitrary* shape cannot really be made to have a unique ID, so we use the pointer instead.

This method leads to good debug output when things go wrong since the shapes are actually printed.